### PR TITLE
feat: swap Preact with React

### PR DESCRIPTION
**Pull request by francoischalifour**
This swaps the Preact implementation with a React implementation.

I had to make the core library events types generic because React uses Synthetic Events. I also fixed some typings issues that Preact types didn't complain about.

React is now a peer dependency of `autocomplete-react`.

### CODE REVIEW

File: packages/autocomplete-core/src/propGetters.ts [hunk 1]
```diff
@@ -50,7 +50,10 @@ export function getPropGetters<TItem>({
         ].some(contextNode => {
           return (
             contextNode &&
-            (isOrContainsNode(contextNode, event.target as Node) ||
+            (isOrContainsNode(
+              contextNode,
+              ((event as unknown) as TouchEvent).target as Node
```
Comment (eunjae-lee) [inline packages/autocomplete-core/src/propGetters.ts]: A genuine question. It cannot be `(event as TouchEvent)`?

Comment (francoischalifour) [reply packages/autocomplete-core/src/propGetters.ts]: For `TouchEvent` we can indeed because I didn't make them generic in the end (they're meant to be attached on the global object).

However for the others we need to cast them first as `unknown` and then as events because e.g. TypeScript cannot know that `TEvent` and `Event` have about the same signature. Passing by `unknown` makes it possible.

There must be a better way to deal with these event types difference but I'm not sure how.

Comment (eunjae-lee) [reply packages/autocomplete-core/src/propGetters.ts]: Ah okay. `as` doesn't force casting. And you had to cast it to `unknown` first to cast it to whatever you want.

**Comment by eunjae-lee** [comment]
Can you ELI5 why you swap Preact with React and if you have any possible approach in mind for vanilla js afterwards?

**Comment by francoischalifour** [comment]
Supporting React by default makes more sense because the primary users of `autocomplete-react` will be React users.

Swapping the implementation to Preact will be straightforward because it's a matter of configuration, as explained in [Preact documentation](https://preactjs.com/guide/v10/getting-started#aliasing-react-to-preact). Before doing so we'll need to understand if we want to reuse this same code.

This moves also makes the environment more stable since React is the main library, and Preact aims at mimicing their public API.

I'm still undecided if we should use Preact for the vanilla version, or plain JavaScript. When we start the vanilla implementation, we should investigate how difficult it would be to write it in plain vanilla. Wild idea, perhaps we should consider [Svelte](https://svelte.dev/) too, which outputs plain JavaScript.

**Comment by eunjae-lee** [comment]
Thanks. Now I'm in the context 🙂
It makes sense and this PR looks good to me.

**Review by eunjae-lee — APPROVED** [review]
